### ISSUE TITLE
Can't find "on_start_epoch, on_sample…" in engine.py

### ISSUE THREAD

**Issue by Huntersxsx** [microsoft/2D-TAN#22]
Hello, thank you for your great work, but I have some questions.
1、In [engine.py](https://github.com/microsoft/2D-TAN/blob/master/lib/core/engine.py), I see some state "on_start_epoch, on_sample, on_end_epoch, on_test_sample", but they only appear once, I can't figure out what's their function, can you explain it?
2、I see you set MAX_EPOCH to 100, but I find the performance on test set has not improved obviously since around 20 epochs, more epochs only improve the performance on training set. Do you have the same situation during your taining time?

**Comment by Sy-Zhang** [comment]
> Hello, thank you for your great work, but I have some questions.
> 1、In [engine.py](https://github.com/microsoft/2D-TAN/blob/master/lib/core/engine.py), I see some state "on_start_epoch, on_sample, on_end_epoch, on_test_sample", but they only appear once, I can't figure out what's their function, can you explain it?
> 2、I see you set MAX_EPOCH to 100, but I find the performance on test set has not improved obviously since around 20 epochs, more epochs only improve the performance on training set. Do you have the same situation during your taining time?

1. "on_start_epoch, on_sample, on_end_epoch, on_test_sample" are not called in our code. We leave it here for debug usage.
2. Yes. We use a large MAX_EPOCH to ensure our model can achieve the best performance under different datasets.

**Comment by Huntersxsx** [comment]
> > Hello, thank you for your great work, but I have some questions.
> > 1、In [engine.py](https://github.com/microsoft/2D-TAN/blob/master/lib/core/engine.py), I see some state "on_start_epoch, on_sample, on_end_epoch, on_test_sample", but they only appear once, I can't figure out what's their function, can you explain it?
> > 2、I see you set MAX_EPOCH to 100, but I find the performance on test set has not improved obviously since around 20 epochs, more epochs only improve the performance on training set. Do you have the same situation during your taining time?
> 
> 1. "on_start_epoch, on_sample, on_end_epoch, on_test_sample" are not called in our code. We leave it here for debug usage.
> 2. For some dataset, the answer is yes. We use a large MAX_EPOCH to ensure our model can achieve the best performance under different datasets.

Thank you for your reply and I have another question to bother you. 
The results you show "Rank1@0.3 | Rank1@0.5 | Rank1@0.7 | Rank5@0.3 | Rank5@0.5 | Rank5@0.7" are from the same checkpoint or you choose the highest score from different checkpoints respectively? 
If the answer is the former, you know there may be no such as a checkpoint reach highest scores on all metrics, (e.g. some checkpoint may have higher Rank1@0.5 score but have lower Rank1@0.7 score), I didn't see you set a rule to stop training in your code, you just train for MAX_EPOCH to get all checkpoints, what's your criterion to choose an appropriate checkpoint to report?

**Comment by Sy-Zhang** [comment]
> > > Hello, thank you for your great work, but I have some questions.
> 
> > > 1、In [engine.py](https://github.com/microsoft/2D-TAN/blob/master/lib/core/engine.py), I see some state "on_start_epoch, on_sample, on_end_epoch, on_test_sample", but they only appear once, I can't figure out what's their function, can you explain it?
> 
> > > 2、I see you set MAX_EPOCH to 100, but I find the performance on test set has not improved obviously since around 20 epochs, more epochs only improve the performance on training set. Do you have the same situation during your taining time?
> 
> > 
> 
> > 1. "on_start_epoch, on_sample, on_end_epoch, on_test_sample" are not called in our code. We leave it here for debug usage.
> 
> > 2. For some dataset, the answer is yes. We use a large MAX_EPOCH to ensure our model can achieve the best performance under different datasets.
> 
> 
> 
> Thank you for your reply and I have another question to bother you. 
> 
> The results you show "Rank1@0.3 | Rank1@0.5 | Rank1@0.7 | Rank5@0.3 | Rank5@0.5 | Rank5@0.7" are from the same checkpoint or you choose the highest score from different checkpoints respectively? 
> 
> If the answer is the former, you know there may be no such as a checkpoint reach highest scores on all metrics, (e.g. some checkpoint may have higher Rank1@0.5 score but have lower Rank1@0.7 score), I didn't see you set a rule to stop training in your code, you just train for MAX_EPOCH to get all checkpoints, what's your criterion to choose an appropriate checkpoint to report? 


Results are from the same checkpoint. For activitynet, we choose the checkpoint with highest rank1@0.3 on validation set. For tacos and charades, we choose the highest rank1@0.5 on testing set, since validation sets are not provided.
